### ISSUE TITLE
mWeb-Conversation-User A in offline typing message, User B is shown "user A is typing"

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#28063]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. As user A, tap profile- Preferences
3. Toggle on forced offline
4. Open 1:1 with user B
5. In other device, as user B, open 1:1 with user A
6. As user A, in offline type messages to user B and note user B conversation page

## Expected Result:
User A in offline typing message, User B must not be shown "user A is typing" in conversation page.

## Actual Result:
User A in offline typing message, User B is shown "user A is typing" in conversation page.

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.73-0

**Reproducible in staging?:** Y

**Reproducible in production?:** No "Force offline" mode

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/a7906296-4aec-4b7e-92d8-10e81afce772

https://github.com/Expensify/App/assets/115492554/2140e236-919b-4574-8b37-9289bdc47122

https://github.com/Expensify/App/assets/115492554/60c3525e-8e82-43b5-8ca2-28fd36060d4f

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause-Internal Team

**Slack conversation:** @

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The user is typing event is sent from a user client even if the client has force offline mode enabled.

### What is the root cause of that problem?
Pusher sends events even if force offline mode is enabled.

### What changes do you think we should make in order to solve the problem?
Modify the Pusher `sendEvent` method to perform an early return if `shouldForceOffline` is true.
```
function sendEvent(channelName, eventName, payload) {
    // Check to see if we are subscribed to this channel before sending the event. Sending client events over channels
    // we are not subscribed too will throw errors and cause reconnection attempts. Subscriptions are not instant and
    // can happen later than we expect.
    if (!isSubscribed(channelName)) {
        return;
    }

    if (shouldForceOffline) {
        Log.info('[Pusher] Ignoring a send event because shouldForceOffline = true');
        return;
    }

   // ... rest of the method
}
```

### What alternative solutions did you explore? (Optional)
None.

**Comment by joelbettner** [comment]
I don't think this is an hourly, since I believe this will only ever happen on staging (I THINK we always hide the TestToolMenu in production...I could be wrong). I don't think this was caused by the recent deploy.  I think this has been happening ever since we introduced the Force Offline mode. @izarutskaya is this something that you just came across, or is this test something that happens at every deploy?

@neil-marcellini @roryabraham it looks like you are the two fellas who have the most background in the Force Offline mode. Care to weigh in here?

**Comment by mountiny** [comment]
Correct, I agree this is not a deploy blocker, also I think @neil-marcellini knows about this problem from before, I believe I remember some slack thread about this specifically.

**Comment by izarutskaya** [comment]
No, we do not  this test at every deploy. This was discovered during another test. Thank you

**Comment by NicMendonca** [comment]
^ setting as external

**Comment by JawadSadiq01** [comment]
### Proposal

**Please re-state the problem that we are trying to solve in this issue.**
Conversation-User A in offline typing message, User B is shown "user A is typing".

**What is the root cause of that problem?**
When **"User A"** types a message, there is no code that checks either he had activated the **Force Offline**. After **Force Offline activation**, when **"User A"** types message, then on each character typing, an event **"debouncedBroadcastUserIsTyping"** triggers. After this event triggering, the **"USER B"** sees that **"USER A"** is typing.

**What changes do you think we should make in order to solve the problem?**
You just need to prevent the call of **"debouncedBroadcastUserIsTyping"**, when he set **Force Offline**.
Just add the below code in **ComposerWithSuggestions.js** file.

Please see the attachments below, for knowing where to add the below code in **ComposerWithSuggestions.js file**.

```
let isForcedOffline = false;

Onyx.connect({
    key: ONYXKEYS.NETWORK,
    callback: (val) => {
        isForcedOffline = val && val.isOffline;
    },
});
```

**On ComposerWithSuggestions.js#216**, put the below code
```
if (newComment && !isForcedOffline) {
   debouncedBroadcastUserIsTyping(reportID);
}
```

<img width="1389" alt="image" src="https://github.com/Expensify/App/assets/135354140/66cf2a42-0796-4344-85a0-09a09b32227b">
<img width="1389" alt="image" src="https://github.com/Expensify/App/assets/135354140/36bdc828-4c70-4616-975f-425937347460">

**Video Recording**
https://drive.google.com/file/d/1G6wSTFZfK1qvSyK2CcB4L4DEIZ_e6WWK/view?usp=drivesdk

**Comment by neil-marcellini** [comment]
I'm happy to take this one over if that's ok with you @joelbettner.

**Comment by neil-marcellini** [comment]
> I'm inclined towards @akinwale's [proposal](https://github.com/Expensify/App/issues/28063#issuecomment-1732297234) (solution 1), because `forced offline` is just a simulated feature in our development/staging environment. Its logic should stay closer to the lower level network side, and not be mixed into our business components. : )
> 
> 🎀 👀 🎀 C+ reviewed

I agree. When I implemented this originally I didn't think about sending data with Pusher, since we hardly ever do that. I'm glad we're cleaning this up.

**Comment by akinwale** [comment]
@ntdiary My [PR](https://github.com/Expensify/App/issues/28063) is ready for review. Thanks.

**Comment by NicMendonca** [comment]
@akinwale @ntdiary bump on BZ check list ^^

**Comment by akinwale** [comment]
> @akinwale @ntdiary bump on BZ check list ^^
> 
> Also can you both apply to the job please? https://www.upwork.com/jobs/~01650cee09674fc251

Applied. Thanks!

I believe it's the C+ that's supposed to complete the BZ check list. Weird that my name is in there. 😅

**Comment by NicMendonca** [comment]
Payment summary: 

Reporter: N/A
Contributor: @akinwale - $750 (paid via upwork)
Contributor +: @ntdiary - $750 (paid via upwork)

**Comment by NicMendonca** [comment]
Sent you both an offer!

**Comment by akinwale** [comment]
> Sent you both an offer!

Accepted. Thanks!
